File: solutions/elixir/bowling/1/lib/bowling.ex
defmodule Bowling do
  @doc """
    Creates a new game of bowling that can be used to store the results of
    the game
  """

  @type strike :: 10
  @type throw :: 0..10
  @type frame :: strike() | {throw(), throw()}
  @type ongoing_frame :: nil | {throw(), nil}
  @type last_frame ::
          ongoing_frame() | frame() | {strike(), strike(), strike()} | {strike(), throw()}
  @type game_state :: %{
          throws: [throw()]
        }

  @spec start() :: game_state()
  def start do
    %{throws: []}
  end

  @doc """
    Records the number of pins knocked down on a single roll. Returns `any`
    unless there is something wrong with the given number of pins, in which
    case it returns a helpful error tuple.
  """

  @spec roll(game_state(), integer) :: {:ok, game_state()} | {:error, String.t()}
  def roll(
        _,
        roll
      )
      when roll > 10 do
    {:error, "Pin count exceeds pins on the lane"}
  end

  def roll(
        _,
        roll
      )
      when roll < 0 do
    {:error, "Negative roll is invalid"}
  end

  def roll(
        %{throws: throws},
        roll
      ) do
    case score_inner(0, 0, Enum.reverse(throws)) do
      {:ongoing, pins} when pins >= roll -> {:ok, %{throws: [roll | throws]}}
      {:ongoing, pins} -> {:error, "Pin count exceeds pins on the lane"}
      {:ok, _} -> {:error, "Cannot roll after game is over"}
    end
  end

  @doc """
    Returns the score of a given game of bowling if the game is complete.
    If the game isn't complete, it returns a helpful error tuple.
  """

  @spec score(game_state()) :: {:ok, integer} | {:error, String.t()}
  def score(game) do
    case score_inner(0, 0, Enum.reverse(game[:throws])) do
      {:ongoing, _} -> {:error, "Score cannot be taken until the end of the game"}
      any -> any
    end
  end

  defp score_inner(acc_score, acc_frames, throws) when acc_frames < 9 do
    case throws do
      [10, y, z | rest] ->
        score_inner(acc_score + 10 + y + z, acc_frames + 1, [y, z | rest])

      [x, y, z | rest] when x + y == 10 ->
        score_inner(acc_score + 10 + z, acc_frames + 1, [z | rest])

      [x, y | rest] ->
        score_inner(acc_score + x + y, acc_frames + 1, rest)

      [10] ->
        {:ongoing, 10}

      [x] ->
        {:ongoing, 10 - x}

      [] ->
        {:ongoing, 10}
    end
  end

  defp score_inner(acc_score, acc_frames, throws) when acc_frames == 9 do
    case throws do
      [10, y, z] ->
        {:ok, acc_score + 10 + y + z}

      [10, 10] ->
        {:ongoing, 10}

      [10, y] ->
        {:ongoing, 10 - y}

      [10] ->
        {:ongoing, 10}

      [x, y, z] when x + y == 10 ->
        {:ok, acc_score + 10 + z}

      [x, y] when x + y == 10 ->
        {:ongoing, 10}

      [x, y] ->
        {:ok, acc_score + x + y}

      [x] ->
        {:ongoing, 10 - x}

      [] ->
        {:ongoing, 10}
    end
  end
end

File: solutions/javascript/bowling/1/bowling.js
//
// For my Shree DR.MDD
//

export class Bowling {
  _rollHistory = [];
  _extraRollAllowed = true;
  _frameCount = 0;
  _pinsInFrame = 0;
  _throwsInFrame = 0;
  _bonusThrows = 0;
  _lastStrike = false;

  roll(pins) {
    this._rollHistory.push(pins);

    if (this._frameCount === 10 && !this._extraRollAllowed) throw new Error('Cannot roll after game is over');

    if (this._frameCount === 10 && !this._lastStrike) {
      this._bonusThrows++;
      if (this._bonusThrows === 2) throw new Error('Cannot roll after game is over');
    } else if (this._frameCount === 10) {
      this._bonusThrows++;
      if (pins < 10 || this._pinsInFrame > 0) this._pinsInFrame += pins;
      if (this._pinsInFrame > 10) throw new Error('Pin count exceeds pins on the lane');
      if (this._bonusThrows === 3) throw new Error('Cannot roll after game is over');
    }

    if (pins < 0) throw new Error('Negative roll is invalid');
    if (pins > 10) throw new Error('Pin count exceeds pins on the lane');

    if (pins === 10 && this._throwsInFrame === 0 && this._frameCount < 10) {
      this._lastStrike = true;
      this._frameCount++;
    }

    if (pins < 10 && this._frameCount < 10) {
      this._pinsInFrame += pins;
      this._throwsInFrame++;

      if (this._pinsInFrame > 10) throw new Error('Pin count exceeds pins on the lane');

      if (this._throwsInFrame === 2) {
        this._frameCount++;
        if (this._frameCount === 10 && this._pinsInFrame === 10) {
          this._extraRollAllowed = true;
          this._lastStrike = false;
        } else if (this._frameCount === 10) {
          this._extraRollAllowed = false;
        }
        this._pinsInFrame = 0;
        this._throwsInFrame = 0;
      }
    }
  }

  score() {
    if (
      this._frameCount < 10 ||
      (this._extraRollAllowed && this._bonusThrows === 0 && !this._lastStrike) ||
      (this._extraRollAllowed && this._bonusThrows < 2 && this._lastStrike)
    ) {
      throw new Error("Score cannot be taken until the end of the game");
    }

    let total = 0;
    let rolls = this._rollHistory;
    let framesLeft = this._frameCount;

    for (let i = 0, frame = 0; frame < 10 && i < rolls.length; frame++) {
      let first = rolls[i];
      let second = rolls[i + 1] ?? 0;
      let third = rolls[i + 2] ?? 0;

      if (first === 10) { // strike
        total += first + second + third;
        i++;
      } else if (first + second === 10) { // spare
        total += first + second + third;
        i += 2;
      } else { // normal frame
        total += first + second;
        i += 2;
      }
    }

    return total;
  }
}
